vsintegration/src/FSharp.ProjectSystem.Base/Project/DocumentManager.cs

```diff
@@ -119,22 +137,131 @@ public virtual int Close(__FRAMECLOSE closeFlag)
+                IVsPersistDocData persistDocData = DocData;
+                if (persistDocData != null)
+                {
+                    string name;
+                    int cancelled;
+                    ErrorHandler.ThrowOnFailure(persistDocData.SaveDocData(VSSAVEFLAGS.VSSAVE_SilentSave, out name, out cancelled));
+                }
             }
         }
 
+        #endregion
+
+        /// <summary>
+        /// Queries the RDT to see if the document is currently edited and not saved.
+        /// </summary>
+        public bool IsDirty
+        {
+            get
+            {
+#if DEV12_OR_LATER
+            var docTable = (IVsRunningDocumentTable4)node.ProjectMgr.GetService(typeof(SVsRunningDocumentTable));
+            if (!docTable.IsMonikerValid(node.GetMkDocument())) {
+                return false;
+            }
+
+            return docTable.IsDocumentDirty(docTable.GetDocumentCookie(node.GetMkDocument()));
+#else
+                bool isOpen, isDirty, isOpenedByUs;
+                uint docCookie;
+                IVsPersistDocData persistDocData;
+                GetDocInfo(out isOpen, out isDirty, out isOpenedByUs, out docCookie, out persistDocData);
+                return isDirty;
+#endif
+            }
+        }
+
+        /// <summary>
+        /// Queries the RDT to see if the document was opened by our project.
+        /// </summary>
+        public bool IsOpenedByUs
+        {
+            get
+            {
+#if DEV12_OR_LATER
+            var docTable = (IVsRunningDocumentTable4)node.ProjectMgr.GetService(typeof(SVsRunningDocumentTable));
+            if (!docTable.IsMonikerValid(node.GetMkDocument())) {
+                return false;
+            }
+
+            IVsHierarchy hierarchy;
+            uint itemId;
+            docTable.GetDocumentHierarchyItem(
+                docTable.GetDocumentCookie(node.GetMkDocument()),
+                out hierarchy,
+                out itemId
+            );
+            return Utilities.IsSameComObject(node.ProjectMgr, hierarchy);
+#else
+                bool isOpen, isDirty, isOpenedByUs;
+                uint docCookie;
+                IVsPersistDocData persistDocData;
+                GetDocInfo(out isOpen, out isDirty, out isOpenedByUs, out docCookie, out persistDocData);
+                return isOpenedByUs;
+#endif
+
+            }
+        }
+
+        /// <summary>
+        /// Returns the doc cookie in the RDT for the associated file.
+        /// </summary>
+        public uint DocCookie
+        {
+            get
+            {
+#if DEV12_OR_LATER
+            var docTable = (IVsRunningDocumentTable4)node.ProjectMgr.GetService(typeof(SVsRunningDocumentTable));
+            if (!docTable.IsMonikerValid(node.GetMkDocument())) {
+                return (uint)ShellConstants.VSDOCCOOKIE_NIL;
+            }
+
+            return docTable.GetDocumentCookie(node.GetMkDocument());
+#else
+                bool isOpen, isDirty, isOpenedByUs;
```

Comment:
Can we remove the conditional and the else block? DEV12_OR_LATER will always be true.